packages/core/contracts/insured-bridge/BridgePool.sol

```diff
@@ -218,6 +218,119 @@ contract BridgePool is Testable, BridgePoolInterface, ExpandedERC20, MultiCaller
+            "Insufficient pool balance"
+        );
+
+        // Compute total proposal bond and pull from caller so that the OptimisticOracle can pull it from here.
+        l1Token.safeTransferFrom(msg.sender, address(this), depositData.amount + relayData.bond);
+        bonds += relayData.bond;
+
+        // Request a price for the relay identifier and propose "true" optimistically. This method will pull the
+        // (proposer reward + proposer bond + final fee) from the caller. We need to set a new price request timestamp
+        // instead of default setting to equal to the `depositTimestamp`, which is dependent on the L2 VM on which the
+        // DepositContract is deployed. Imagine if the timestamps on the L2 have an offset that are always "in the
+        // future" relative to L1 blocks, then the OptimisticOracle would always reject requests.
+        // _requestAndProposeOraclePriceRelay(proposerBond, priceRequestTime, ancillaryData);
+
+        pendingReserves += depositData.amount; // Book off maximum liquidity used by this relay in the pending reserves.
+
+        bytes32 instantRelayHash = _getInstantRelayHash(depositHash, relayData);
+        require(
+            // Can only speed up a pending relay without an existing instant relay associated with it.
+            instantRelays[instantRelayHash] == address(0),
+            "Relay cannot be sped up"
+        );
+        instantRelays[instantRelayHash] = msg.sender;
+
+        // Pull relay amount minus fees from caller and send to the deposit l1Recipient. The total fees paid is the sum
+        // of the LP fees, the relayer fees and the instant relay fee.
+        uint256 feesTotal =
+            _getAmountFromPct(
+                relayData.realizedLpFeePct + depositData.slowRelayFeePct + depositData.instantRelayFeePct,
+                depositData.amount
+            );
+        // If the L1 token is WETH then: a) pull WETH from instant relayer b) unwrap WETH c) send ETH to recipient.
+        uint256 recipientAmount = depositData.amount - feesTotal;
+        if (isWethPool) {
+            WETH9Like(address(l1Token)).withdraw(recipientAmount);
+            depositData.l1Recipient.transfer(recipientAmount);
+        }
+        // Else, this is a normal ERC20 token. Send to recipient.
+        else l1Token.safeTransfer(depositData.l1Recipient, recipientAmount);
+
+        emit DepositRelayed(depositHash, depositData, address(l1Token), relayData, relayHash);
+        emit RelaySpedUp(depositHash, msg.sender, relayData);
+    }
+
+    function disputeRelay(DepositData memory depositData, RelayData memory relayData) public nonReentrant() {
+        // Validate the input data.
+        bytes32 depositHash = _getDepositHash(depositData);
+        _validateRelayDataHash(depositHash, relayData);
+
+        // Submit the proposal and dispute to the OO.
+        bytes32 relayHash = _getRelayHash(depositData, relayData);
+        _requestProposeDispute(
```

Comment:
Another minor problem is what happens if the collateral token stops being whitelisted or similar. Instead of just blocking relays, this would block only disputes.

One idea is to try-catch here and automatically pause settlements if a dispute fails. Then unpause once one succeeds.

Reply:
> One idea is to try-catch here and automatically pause settlements if a dispute fails. Then unpause once one succeeds.

This is useful because if we pause it, then we can re-whitelist the collateral, and then unpause it?

Reply:
Exactly. The only issue is that it requires someone to issue a dispute to unpause :/, which has some weird incentives now that I think about it.